homeassistant/components/schedule/__init__.py

```diff
@@ -109,16 +110,21 @@ def serialize_to_time(value: Any) -> Any:
     vol.Optional(CONF_ICON): cv.icon,
 }
 
+# Extra data that the user can set on each time range
+CUSTOM_DATA_SCHEMA = vol.Or(vol.Schema({str: vol.Coerce(str)}), vol.Coerce(str))
+
 TIME_RANGE_SCHEMA = {
     vol.Required(CONF_FROM): cv.time,
     vol.Required(CONF_TO): deserialize_to_time,
+    vol.Optional(CONF_DATA): CUSTOM_DATA_SCHEMA,
 }
 
 # Serialize time in validated config
 STORAGE_TIME_RANGE_SCHEMA = vol.Schema(
     {
         vol.Required(CONF_FROM): vol.Coerce(str),
         vol.Required(CONF_TO): serialize_to_time,
+        vol.Optional(CONF_DATA): CUSTOM_DATA_SCHEMA,
```

Comment:
- The schema can be inlined, no need to create a constant in this case. 
- Any reason this is limited/coerced to strings?

Reply:
1. I created a constant because it's used twice, once in `TIME_RANGE_SCHEMA` and again in `STORAGE_TIME_RANGE_SCHEMA`. Should I inline it even though it will be repeated?
2. I coerced everything to strings because I wasn't sure how to represent "any scalar". Using `typing.Any` would theoretically allow nested dicts and lists (assuming the attribute code can handle that) but voluptuous didn't want to validate anything when I used it. Would `vol.Or(bool, str, int, float)` be more appropriate?

Reply:
1. if it was a huge schema used many times, I would agree; this is overkill imho.
2. your suggestion seems acceptable

Reply:
I replaced the string coercion, so now it's
```py
CUSTOM_DATA_VALUE = vol.Or(bool, str, int, float)
CUSTOM_DATA_SCHEMA = vol.Or(vol.Schema({str: CUSTOM_DATA_VALUE}), CUSTOM_DATA_VALUE)
```

Expanding this would give:
```py
TIME_RANGE_SCHEMA: VolDictType = {
    vol.Required(CONF_FROM): cv.time,
    vol.Required(CONF_TO): deserialize_to_time,
    vol.Optional(CONF_DATA): vol.Or(vol.Schema({str: vol.Or(bool, str, int, float)}), vol.Or(bool, str, int, float)),
}

STORAGE_TIME_RANGE_SCHEMA = vol.Schema(
    {
        vol.Required(CONF_FROM): vol.Coerce(str),
        vol.Required(CONF_TO): serialize_to_time,
        vol.Optional(CONF_DATA): vol.Or(vol.Schema({str: vol.Or(bool, str, int, float)}), vol.Or(bool, str, int, float)),
    }
)
```

This feels redundant and more fragile to maintain, as it's not clear that these have elements that need to match each other. If you'd still prefer this I can change it, but I think it's worth avoiding the repetition.

Reply:
I agree it's better to avoid repeating 👍